packages/core/src/indexing/insertion-order.ts

```diff
@@ -128,72 +165,116 @@ export class InsertionOrder {
+      pagination.first + 1
+    )
   }
 
   /**
    * Backward query: traverse from the last to the most recent.
    */
-  private backwardQuery(
+  private async backwardQuery(
     query: BaseQuery,
     pagination: BackwardPaginationQuery
-  ): Knex.QueryBuilder<unknown, Array<Selected>> {
-    const tableName = asTableName(query.model)
-    const queryFunc = this.query(query, true)
-    return this.dbConnection
-      .select('*')
-      .from((bldr) => {
-        let subquery = queryFunc(bldr.from(tableName)).limit(pagination.last + 1)
-        if (pagination.before) {
-          const before = Cursor.parse(pagination.before)
-          subquery = subquery.where('created_at', '<', before.created_at)
-        }
-        return subquery.as('T')
-      })
-      .orderBy(INSERTION_ORDER)
+  ): Promise<Array<QueryResult>> {
+    const response = await this.query(query, true, Cursor.parse(pagination.before)).limit(
+      pagination.last + 1
+    )
+    // Reverse response as results are returned in descending order
+    response.reverse()
+    return response
   }
 
-  private query(query: BaseQuery, isReverseOrder: boolean): QueryFunc {
-    let converted = null
-    if (query.queryFilters) {
-      const parsed = parseQueryFilters(query.queryFilters)
-      converted = convertQueryFilter(parsed)
+  private query(query: BaseQuery, isReverseOrder: boolean, cursor?: CursorData): QueryBuilder {
+    let builder: QueryBuilder = this.dbConnection
+      .from(asTableName(query.model))
+      .columns(['stream_id', 'last_anchored_at', 'created_at', DATA_FIELD])
+      .select()
+    // Handle filters (account, fields and/or legacy relations)
+    builder = this.applyFilters(builder, query)
+    const sorting = query.sorting ?? {}
+    // Handle cursor if present
+    if (cursor != null) {
+      builder = this.applyCursor(builder, cursor, isReverseOrder, sorting)
     }
-    return (bldr) => {
-      let base = bldr.columns(['stream_id', 'last_anchored_at', 'created_at', DATA_FIELD]).select()
+    // Handle ordering
+    builder = this.applySorting(builder, isReverseOrder, sorting)
+    return builder
+  }
 
-      if (converted) {
-        base = base.where(converted.where)
-      }
+  private applyFilters(builder: QueryBuilder, query: BaseQuery): QueryBuilder {
+    if (query.account) {
+      builder = builder.where({ controller_did: query.account })
+    }
 
-      if (isReverseOrder) {
-        base = base.orderBy(reverseOrder(INSERTION_ORDER))
-      } else {
-        base = base.orderBy(INSERTION_ORDER)
+    if (query.queryFilters) {
+      const parsed = parseQueryFilters(query.queryFilters)
+      const converted = convertQueryFilter(parsed)
+      if (converted) {
+        builder = builder.where(converted.where)
       }
-      if (query.account) {
-        base = base.where({ controller_did: query.account })
+    } else if (query.filter) {
+      // Handle legacy `filter` object used for relations
```

Comment:
what if you query both on a relation and a custom indexed field?  Does it all go into queryFilters?  If so, could we just use the existing `filter` arg for both?  Is there a reason we need to introduce a new arg here?

Reply:
With v0.5 ComposeDB uses only `queryFilters` for both relations and custom fields.
We can't use the existing `filter` arg without breaking backwards compatibility, but we should remove it in the next major version of Ceramic.

Reply:
why can't we re-use filter?  Is the format of how the args within it are specified so different?

Reply:
Yes, cf the type definitions here: https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/common/src/index-api.ts#L116-L117

Reply:
I see.  So are queries on relations expressed as `where` queries now?

Reply:
Yes currently they are expressed as `where` queries with exact values matches, as it's what we needed for relations (matching the wanted stream ID)